streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalSegmentIterator.java

```
@@ -0,0 +1,138 @@
+import org.apache.kafka.streams.state.VersionedRecord;
+import org.apache.kafka.streams.state.VersionedRecordIterator;
+import org.rocksdb.Snapshot;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class LogicalSegmentIterator implements VersionedRecordIterator {
+    protected final ListIterator<LogicalKeyValueSegment> segmentIterator;
+    private final Bytes key;
+    private final Long fromTime;
+    private final Long toTime;
+    private final ResultOrder order;
+    protected ListIterator<VersionedRecord<byte[]>> iterator;
+
+    // defined for creating/releasing the snapshot. 
+    private LogicalKeyValueSegment snapshotOwner;
+    private Snapshot snapshot;
+
+
+
+    public LogicalSegmentIterator(final ListIterator<LogicalKeyValueSegment> segmentIterator,
+                                  final Bytes key,
+                                  final Long fromTime,
+                                  final Long toTime,
+                                  final ResultOrder order) {
+
+        this.segmentIterator = segmentIterator;
+        this.key = key;
+        this.fromTime = fromTime;
+        this.toTime = toTime;
+        this.iterator = Collections.emptyListIterator();
+        this.order = order;
+        this.snapshot = null;
+        this.snapshotOwner = null;
+    }
+
+    @Override
+    public void close() {
+        // user may refuse consuming all returned records, so release the snapshot when closing the iterator if it is not released yet!
+        releaseSnapshot();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return iterator.hasNext() || maybeFillIterator();
+    }
+
+    @Override
+    public Object next() {
+        if (hasNext()) {
+            return iterator.next();
+        }
+        return null;
```

I think we should throw an exception for this case (or maybe we can just inherit from `AbstractIterator` instead.